lib/configuration.ex

```diff
@@ -0,0 +1,102 @@
+defmodule Telepoison.Configuration do
+  @moduledoc false
+
+  @default_route_inference_function &Telepoison.URI.infer_route_from_request/1
+
+  defstruct [:infer_route, :ot_attributes]
+
+  alias HTTPoison.Request
+
+  @spec setup(infer_route: (Request.t() -> String.t()), ot_attributes: [{String.t(), String.t()}]) :: :ok
+  def setup(opts \\ []) do
+    Agent.start_link(fn -> set_configuration(opts) end, name: __MODULE__)
```

Comment:
> Are we sure that using an Agent here is the best way to do this? Wouldn't an application env keyed by the module name make more sense?

This is honestly what I wanted to originally suggest, but didn't mean to be too rude :laughing: 
It seems it's the more 'erlang' way to do it tbh

Reply:
I completely forgot about that pattern :facepalm: Yes, it would be more natural, I'll implement it that way

Reply:
https://hexdocs.pm/elixir/1.13/Application.html#c:start/2 <- this is were you should look at (IMO) for this kind of stuff; if it's an application (doing things at startup) we should treat it as such. This is what I meant with the other comment, that, to be honest, is poorly written -.-"
Talking about
> this is an application (of sort, more like a shallow application used to start and configure a collection of applications)

Reply:
More on that: as far as I remember elixir adds every dep to the required applications list when generating an erlang release so you wouldn't even need to actively call `Telepoison.setup` if you'd implement a proper `Application.start/2` function. Can't remember why it wasn't implemented like that tbh :shrug:

Reply:
I think it was implemented like this because `Telepoison` technically doesn't need to be set up, it's just a wrapper around HTTPoison. Also `setup` would be better called `set_default_options` or `configure` or something like that

If I understand correctly there are (at least) four ways to handle "configuring" something in Elixir

1. Define a macro that just substitutes those values e.g. bridge_ex -> not applicable here
2. Fetch stuff from `Application` env -> possibly risky if `Telepoison` is used in umbrella apps and each app wants their own config
3. Use an `Agent` like we are doing here -> I think it incurs the same problem from **2** since the agent would have the same name for each umbrella app
4. Leverage `Application.start` to define startup values -> `Telepoison` shouldn't need to be started up this way because it should be just a wrapper

Did I understand correctly?

Reply:
> Telepoison shouldn't need to be started up this way because it should be just a wrapper

Funny thing is elixir will do it anyway, as it does for every dependency you add; they get added to the erlang release as the main application dependencies, so they get started before your app anyway

> possibly risky if Telepoison is used in umbrella apps and each app wants their own config

I think that is true for every approach you mentioned and it is not how umbrella apps work anyway (they shared the conf anyway); point 4 can be implemented via point 2 or 3 btw :D

Reply:
Ok! I'll try to do 2 and 4 then :smile: